onnxruntime/test/python/onnxruntime_test_python.py

```diff
@@ -641,6 +642,47 @@ def testSessionOptionsAddFreeDimensionOverrideByName(self):
         # "Dim1" and "Dim2" have values assigned to them.
         self.assertEqual(input_shape, [4, 6, 5])
 
+    def testSessionOptionsRegisterCustomOpsLibrary(self):
+        this = os.path.dirname(__file__)
+        if sys.platform.startswith("win"):
+            shared_library = os.path.join(this, "custom_op_library.dll")
+            if not os.path.exists(shared_library):
+                raise FileNotFoundError("Unable to find '{0}'".format(shared_library))
+
+        elif sys.platform.startswith("darwin"):
+            shared_library = os.path.join(this, "libcustom_op_library.dylib")
+            if not os.path.exists(shared_library):
+                raise FileNotFoundError("Unable to find '{0}'".format(shared_library))
+
+        else:
+            shared_library = os.path.join(this, "libcustom_op_library.so")
+            if not os.path.exists(shared_library):
+                raise FileNotFoundError("Unable to find '{0}'".format(shared_library))
+
+        custom_op_model = os.path.join(this, "testdata", "custom_op_library", "custom_op_test.onnx")
+        if not os.path.exists(custom_op_model):
+            raise FileNotFoundError("Unable to find '{0}'".format(custom_op_model))
+
+        so1 = onnxrt.SessionOptions()
+        so1.register_custom_ops_library(shared_library)
+
+        # Create an alias for SessionOptions instance
+        so2 = so1
+
+        # Model loading successfully indicates that the custom op node could be resolved successfully
+        sess1 = onnxrt.InferenceSession(custom_op_model, so1)
```

Comment:
can we add custom_op_lib_paths as an optional arguments? So the user doesn't need to jump to another sessionOption to do that. And this looks more pythonic.

Reply:
Do you want to support an arg in InferenceSession ? That is difficult to do.

Also, the user doesn't need to jump to another session options. There is only one session options.

```
so = ort.SessionOptions()
so.register_custom_ops_library("path")
sess = ort.InferenceSession(model_path, so)
```

I think this is very simple and it aligns with all the other language bindings. Also, session options can be shared across multiple sessions. If we tie it to Session, we have to keep referencing the library in all the Sessions.

Reply:
I added another session options just to demonstrate that it is "re-useable"

Reply:
It may not be difficult as you think. 
In InferenceSession.__init__ python function, just add a few lines as following
  if custom_op_lib_paths:
     so.register_custom_ops_library(custom_op_lib_paths)
  ...

The overall C++ and pybinding interface keeps same, the only difference is to make python wrapper more friendly to avoid everyone write couple of same code lines.

Reply:
`if custom_op_lib_paths:
so.register_custom_ops_library(custom_op_lib_paths)`

I find this hacky and not providing too much value (saves 2 lines of code while potentially causing maintenance hell) for several reasons:

* What is `so` here then ? Construct a SessionOptions instance in InferenceSession on the fly just to hold the custom op libs ? 

* What if the user provides his/her own SessionOptions ? Then do we add the custom_op to that SessionOptions instance (and mutate something the user provided) or do we hold another "dummy" SessionOptions instance just to hold the custom_ops ?

* I think this is unnecessary complications just to save `so = Ort.SessionOptions so.register_ops_library(path)`. This is more clear in that it ties the shared library to the SessionOptions instance and thereby allowing a user to share it across sessions if he/she chooses to do so. 

* Also, by exposing the extra optional parameter in InferenceSession init, it does not keep it consistent with c/c++/c# - none of these language bindings will expose a Session constructor that takes a pth to the shared library - they need to pass it in via SessionOptions as well. My suggestion is to stick with the same.

Is there any concern with instantiating a SessionOptions instance ?

Reply:
Unless you think these two lines are problematic, I don't see what's the difference between putting it in ort code wrapper and the user code, right? And if you are the user, will it be helpful if there is a simpler API for you?

And No one expects C++ and Python interface are exactly same, only the parity features.

Is there any ETA on merging this PR? My concern is not a blocking issue here.

Reply:
Thanks Wenbing.I think it is problematic for the concerns/ reasons listed above. Also, I don’t want to set the precedence of exposing an optional parameter in the InferenceSession constructor for every Session option because that defeats the purpose of SessionOptions.

The PR will be merged once I solve all build issues- working on it now.